File: src/lib/region-evidence-rules.ts
import rulesData from "../../data/region-evidence-rules.json";
import type { ListingRegionEvidence } from "./listing-region-verification";

type RegionRuleBlock = {
  requiredAnyOf?: string[];
  minEvidenceCount?: number;
  forbiddenEvidence?: string[];
  minAiConfidence?: number;
  summary?: string;
};

type RulesFile = {
  default: RegionRuleBlock;
  catalogRegionOverrides: Record<string, RegionRuleBlock>;
  platforms: Record<string, RegionRuleBlock & { catalogRegionOverrides?: Record<string, RegionRuleBlock> }>;
  evidenceLabels: Record<string, string>;
};

const rules = rulesData as RulesFile;

function normalizeRegion(region: string): string {
  return region.trim().toLowerCase();
}

function mergeRule(...blocks: (RegionRuleBlock | undefined)[]): RegionRuleBlock {
  const merged: RegionRuleBlock = {};
  for (const block of blocks) {
    if (!block) continue;
    if (block.requiredAnyOf) merged.requiredAnyOf = block.requiredAnyOf;
    if (block.minEvidenceCount != null) merged.minEvidenceCount = block.minEvidenceCount;
    if (block.forbiddenEvidence) merged.forbiddenEvidence = block.forbiddenEvidence;
    if (block.minAiConfidence != null) merged.minAiConfidence = block.minAiConfidence;
    if (block.summary) merged.summary = block.summary;
  }
  return merged;
}

/** Reglas efectivas para un título concreto (plataforma + región catálogo). */
export function getRegionEvidenceRule(platformSlug: string, catalogRegion: string): RegionRuleBlock {
  const regionKey = normalizeRegion(catalogRegion);
  const platform = rules.platforms[platformSlug];

  return mergeRule(
    rules.default,
    rules.catalogRegionOverrides[regionKey],
    platform,
    platform?.catalogRegionOverrides?.[regionKey],
  );
}

export function getEvidenceLabel(code: string): string {
  return rules.evidenceLabels[code] ?? code;
}

export type EvidenceCheckResult =
  | { ok: true }
  | { ok: false; reason: "insufficient_count" | "missing_required" | "forbidden" | "low_ai_confidence" };

export function checkListingEvidenceMeetsRules(input: {
  platformSlug: string;
  catalogRegion: string;
  regionEvidence: string[];
  aiConfidence?: number | null;
}): EvidenceCheckResult {
  const rule = getRegionEvidenceRule(input.platformSlug, input.catalogRegion);
  const evidence = new Set(input.regionEvidence.map((e) => e.trim()).filter(Boolean));

  const minCount = rule.minEvidenceCount ?? rules.default.minEvidenceCount ?? 1;
  if (evidence.size < minCount) {
    return { ok: false, reason: "insufficient_count" };
  }

  const requiredAny = rule.requiredAnyOf ?? [];
  if (requiredAny.length > 0 && !requiredAny.some((code) => evidence.has(code))) {
    return { ok: false, reason: "missing_required" };
  }

  const forbidden = rule.forbiddenEvidence ?? [];
  if (forbidden.some((code) => evidence.has(code))) {
    return { ok: false, reason: "forbidden" };
  }

  const minAi = rule.minAiConfidence ?? rules.default.minAiConfidence;
  if (
    minAi != null &&
    input.aiConfidence != null &&
    input.aiConfidence < minAi
  ) {
    return { ok: false, reason: "low_ai_confidence" };
  }

  return { ok: true };
}

export function formatRegionEvidenceRuleSummary(platformSlug: string, catalogRegion: string): string {
  const rule = getRegionEvidenceRule(platformSlug, catalogRegion);
  if (rule.summary) return rule.summary;

  const required = rule.requiredAnyOf ?? [];
  if (required.length === 0) {
    return `Al menos ${rule.minEvidenceCount ?? 1} prueba de región verificada.`;
  }

  const labels = required.map((code) => getEvidenceLabel(code));
  return `Al menos una de: ${labels.join("; ")}.`;
}

export function listRequiredEvidenceCodes(
  platformSlug: string,
  catalogRegion: string,
): ListingRegionEvidence[] {
  const rule = getRegionEvidenceRule(platformSlug, catalogRegion);
  return (rule.requiredAnyOf ?? []) as ListingRegionEvidence[];
}

File: scripts/region_evidence_rules.py
"""Reglas de prueba de región (data/region-evidence-rules.json)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
RULES_FILE = ROOT / "data" / "region-evidence-rules.json"


def _load_rules() -> dict[str, Any]:
    return json.loads(RULES_FILE.read_text(encoding="utf-8"))


def _normalize_region(region: str) -> str:
    return region.strip().lower()


def _merge_rule(*blocks: dict[str, Any] | None) -> dict[str, Any]:
    merged: dict[str, Any] = {}
    for block in blocks:
        if not block:
            continue
        for key, value in block.items():
            if value is not None:
                merged[key] = value
    return merged


def get_region_evidence_rule(platform_slug: str, catalog_region: str) -> dict[str, Any]:
    rules = _load_rules()
    region_key = _normalize_region(catalog_region)
    platform = (rules.get("platforms") or {}).get(platform_slug) or {}

    return _merge_rule(
        rules.get("default"),
        (rules.get("catalogRegionOverrides") or {}).get(region_key),
        platform,
        (platform.get("catalogRegionOverrides") or {}).get(region_key),
    )


def get_retail_evidence_rule(retail_source: str) -> dict[str, Any]:
    rules = _load_rules()
    return dict((rules.get("retailSources") or {}).get(retail_source) or {})


def check_retail_evidence_meets_rules(
    retail_source: str,
    region_evidence: list[str],
    ai_confidence: float | None = None,
) -> tuple[bool, str | None]:
    rules = _load_rules()
    rule = get_retail_evidence_rule(retail_source)
    if not rule:
        return False, "unknown_retail_source"

    evidence = {e.strip() for e in region_evidence if e and str(e).strip()}
    min_count = rule.get("minEvidenceCount") or rules.get("default", {}).get("minEvidenceCount", 1)
    if len(evidence) < min_count:
        return False, "insufficient_count"

    required_any = rule.get("requiredAnyOf") or []
    if required_any and not any(code in evidence for code in required_any):
        return False, "missing_required"

    forbidden = rule.get("forbiddenEvidence") or []
    if any(code in evidence for code in forbidden):
        return False, "forbidden"

    min_ai = rule.get("minAiConfidence") or rules.get("default", {}).get("minAiConfidence")
    if min_ai is not None and ai_confidence is not None and ai_confidence < min_ai:
        return False, "low_ai_confidence"

    return True, None


def check_listing_evidence_meets_rules(
    platform_slug: str,
    catalog_region: str,
    region_evidence: list[str],
    ai_confidence: float | None = None,
) -> tuple[bool, str | None]:
    rules = _load_rules()
    rule = get_region_evidence_rule(platform_slug, catalog_region)
    evidence = {e.strip() for e in region_evidence if e and str(e).strip()}

    min_count = rule.get("minEvidenceCount") or rules.get("default", {}).get("minEvidenceCount", 1)
    if len(evidence) < min_count:
        return False, "insufficient_count"

    required_any = rule.get("requiredAnyOf") or []
    if required_any and not any(code in evidence for code in required_any):
        return False, "missing_required"

    forbidden = rule.get("forbiddenEvidence") or []
    if any(code in evidence for code in forbidden):
        return False, "forbidden"

    min_ai = rule.get("minAiConfidence") or rules.get("default", {}).get("minAiConfidence")
    if min_ai is not None and ai_confidence is not None and ai_confidence < min_ai:
        return False, "low_ai_confidence"

    return True, None
